core/src/credential.rs

```diff
@@ -0,0 +1,93 @@
+
+#![allow(unused_variables)]
+#![allow(dead_code)]
+use std::{time::{SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use crate::HashMap;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VerificationContext = [&'static str; 2];
+
+pub const CONTEXT_CREDENTIALS: VerificationContext = [
+    "https://www.w3.org/2018/credentials/v1",
+    "https://www.w3.org/2018/credentials/examples/v1",
+];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: &'static str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: &'static str = "BankCard";
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+struct CredentialSubject {
+    id : String,
+    #[serde(flatten)]
+    pub property_set: HashMap<String, Value>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(bound(deserialize = "'de: 'static"))]
+pub struct Credential {
```

Comment:
Same issue as mentioned before with this struct and its relationship to `VerifiableCredential`.

A few things:
1. A "Credential" will not have a proof as its not issued
2. I think the Credential should have its own properties and we shouldn't tie them in this way.

Reply:
What if we return Credential, and will have a separate type for the signed one, like "VerifiedCredential"?

Reply:
And the relationship between them will be like VerifiedCredential contains Credential and all Credential fields will be flattened in VerifiedCredential? The idea behind it is that there is no point in maintaining a separate structure for the unsigned and signed credential, since they are essentially the same.

Reply:
Yeah a VerifiableCredential = Credential + Proof. I agree that it's not great to have to support both. But using the serde flattening feels like a hack around the limitations of the union type. I don't know what the Rust standard is for these situations. @mkatychev?

Rust type unions require unsafe reading which is why I think that not using them is a good idea. But I'm curious as to what the recommended Rust solution is here. Apart from maintaining two structs which is definitely unideal but seems to me better than relying on serde to flatten the definitions and hence doing a runtime union

Reply:
@sKudryashov @mkatychev If there is no way of doing something like a Union then let's just keep it simple and keep the flattening. So the only changes I'd need to see is:

1. create_credential should return a `Credential`
2. The relationship should be flipped s.t a VerifiableCredential contains the Credential and a Proof

Reply:
@LuisOsta  unless this PR needs to be closed to address your concerns, let's create a ticket  or open a discussion and merge this one. There's too many disparate discussions in this PR and too much to discuss in general.

Reply:
@mkatychev I agree that this has gone too long but points 1 and 2 raised in my last comment are easy changes and fundamental to this PR (i.e. it'll be non functional if merged)

Reply:
Yeah they both sound like reasonable changes. I'm not sure if `VerfiableCredential` is causal to `Credential` or the other way around, I suppose that conclusion would guide me towards 1 or 2.